### PR TITLE
Split species

### PR DESCRIPTION
**Pull request by marwahaha**

### CODE REVIEW

File: src/IonSim.jl [hunk 1]
```diff
@@ -25,6 +25,7 @@ include("traps.jl")
 include("operators.jl")                           
 include("hamiltonians.jl")
 include("time_evolution.jl")
+include("species/include_species.jl")
```
Comment (marwahaha) [inline src/IonSim.jl:28]: new folder on species

make a new file for every ion species.

File: src/constants.jl [hunk 2]
```diff
@@ -14,12 +14,29 @@ struct PhysicalConstant <: Real
...
+# """ ## `m_hg198` = 3.2873155315e-25 <br> (mass of 198Hg)"""
+# const m_hg198 = PhysicalConstant(3.2873155315e-25, "kg")
+# """ ## `m_hg199` = 3.3039460302e-25 <br> (mass of 199Hg)"""
+# const m_hg199 = PhysicalConstant(3.3039460302e-25, "kg")
+
 """ ## `μB` = 9.27400994e-24 J⋅T⁻¹ <br> (Bohr Magneton)"""
 const μB = PhysicalConstant(9.27400994e-24, "J⋅T⁻¹")
 """ ## `ħ` = 1.0545718e-34 m²kg/s <br> (Planck's constant / 2π)"""
 const ħ = PhysicalConstant(1.0545718e-34, "m²kg/s")
-""" ## `m_ca40` = 6.635943757345042e-26 kg <br> (mass of 40Ca)"""
```
Comment (marwahaha) [inline src/constants.jl:21]: these get moved to species/ folder.

File: src/hamiltonians.jl [hunk 3]
```diff
@@ -1,4 +1,4 @@
-using SparseArrays: rowvals, nzrange, nonzeros
+using SparseArrays: rowvals, nzrange, spzeros, spdiagm, findnz, rowvals, nonzeros
```
Comment (marwahaha) [inline src/hamiltonians.jl:1]: joe made some recent changes; maybe neil could look at this?

File: src/ions.jl [hunk 4]
```diff
@@ -1,10 +1,14 @@
 using WignerSymbols: wigner3j
-using .PhysicalConstants: e, ca40_qubit_transition_frequency, m_ca40, ħ, α, μB
+using .PhysicalConstants: e, ħ, α, μB
```
Comment (marwahaha) [inline src/ions.jl:2]: This is the biggest changed file.

File: src/ions.jl [hunk 5]
```diff
@@ -17,212 +21,164 @@ The physical parameters defining an isolated ion's internal structure.
...
+
+
+""" 
+This needs a docstring
 """
-    Ca40(selected_level_structure::Vector{String}[, stark_shift])
-
-#### user-defined fields
-* `selected_level_structure`: 
-    keys ⊂ `["S-1/2", "S+1/2", "D-5/2", "D-3/2", "D-1/2", "D+1/2", "D+3/2", "D+5/2"]`.
```
Comment (marwahaha) [inline src/ions.jl:40]: This is easy mode. For Ca-43 .... there are ~100 different sublevels!

File: src/traps.jl [hunk 6]
```diff
@@ -345,7 +351,7 @@ function set_gradient!(
     ion1 = T.configuration.ions[ion_indxs[1]]
     ion2 = T.configuration.ions[ion_indxs[2]]
     separation = abs(ion1.position - ion2.position)
-    E1, E2 = map(x -> zeeman_shift(1, ion1.selected_level_structure[x]), transition)
+    E1, E2 = map(x -> zeeman_shift(1, sublevel_structure(ion1, x)), transition)
```
Comment (marwahaha) [inline src/traps.jl:354]: no tests yet!

File: src/traps.jl [hunk 7]
```diff
@@ -311,26 +310,33 @@ transition_frequency(
...
     )
     ion = T.configuration.ions[ion_index]
     B = T.B + T.∇B * ion.position
     transition_frequency(B, ion, transition)
 end
 
+"""
+This needs a docstring.
+"""
+matrix_element(I::Ion, transition::Tuple, T::Trap, laser::Laser, time::Real) = matrix_element(I, transition, laser.E(time), laser.k, laser.ϵ, T.Bhat)
```
Comment (marwahaha) [inline src/traps.jl:336]: This code is not written yet

File: src/IonSim.jl [hunk 8]
```diff
@@ -25,6 +25,7 @@ include("traps.jl")
 include("operators.jl")                           
 include("hamiltonians.jl")
 include("time_evolution.jl")
```
Comment (marwahaha) [inline src/IonSim.jl:27]: This is not compatible with old versions of ionsim!

Comment (marwahaha) [reply src/IonSim.jl:27]: whenever i instantiate an ion, it will have to change.

File: src/constants.jl [hunk 9]
```diff
@@ -14,12 +14,29 @@ struct PhysicalConstant <: Real
 end
 
 # some useful constants, everything in SI units
```
Comment (marwahaha) [inline src/constants.jl:16]: need to add documentation

File: src/constants.jl [hunk 10]
```diff
@@ -82,4 +97,4 @@ function Base.:-(a::NamedTuple{(:x,:y,:z)}, b::NamedTuple{(:x,:y,:z)})
 end
 Base.:/(a::NamedTuple{(:x,:y,:z)}, b::Number) = (x=a.x/b, y=a.y/b, z=a.z/b)
 Base.:*(a::NamedTuple{(:x,:y,:z)}, b::Number) = (x=a.x*b, y=a.y*b, z=a.z*b)
-Base.:*(b::Number, a::NamedTuple{(:x,:y,:z)}) = (x=a.x*b, y=a.y*b, z=a.z*b)
```
Comment (marwahaha) [inline src/constants.jl:85]: need to test!

File: src/lasers.jl [hunk 11]
```diff
@@ -22,21 +22,22 @@ The physical parameters defining laser light.
...
     Δ::Real
     ϵ::NamedTuple{(:x,:y,:z)}
     k::NamedTuple{(:x,:y,:z)}
     ϕ::Function 
-    λ::Real
     pointing::Vector
-    function Laser(; 
-            E::TE=0, Δ=0, ϵ=(x̂+ŷ)/√2, k=ẑ, ϕ::Tϕ=0, λ=c/ca40_qubit_transition_frequency, 
+    function Laser(;
+            λ = nothing, E::TE=0, Δ=0, ϵ=(x̂+ŷ)/√2, k=ẑ, ϕ::Tϕ=0,
```
Comment (marwahaha) [inline src/lasers.jl:33]: why is $\lambda$ here?

File: src/species/ca40.jl [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+                         # Optional fields
+                         default_sublevel_selection = [("S1/2", "all"),
+                                                       ("D5/2", "all"),
+                                                      ],
+
+                         gfactors = Dict("S1/2" => 2.00225664,
+                                         "D5/2" => 1.2003340
+                                        ),
+
+                         #nonlinear_zeeman = Dict(("S1/2", -1//2) => B->1.3e-4*B^2,
```
Comment (marwahaha) [inline src/species/ca40.jl:36]: Do we need this?

File: src/traps.jl [hunk 13]
```diff
@@ -183,14 +192,13 @@ transition)`
...
+    transition::Tuple
 )   
     p = laser.pointing
-    (γ, ϕ) = map(x -> rad2deg(acos(ndot(Bhat, x))), [laser.ϵ, laser.k])
-    s_indx = findall(x -> x[1] == ion.number, p)
+    s_indx = findall(x -> x[1] == ionnumber(ion), p)
     @assert length(s_indx) > 0 "This laser doesn't shine on this ion"
-    s = laser.pointing[s_indx[1]][2]
-    Ω = s * ion.selected_matrix_elements[tuple(transition...)](1, γ, ϕ)
+    s = p[s_indx[1]][2]
```
Comment (marwahaha) [inline src/traps.jl:200]: what is p?

File: src/species/ca40.jl (hunk 12 above)
Comment (neil-glikin) [reply src/species/ca40.jl:36]: The new code allows for assigning an ion species an optional nonlinear term of the Zeeman shift (relevant in some use cases). We have no need for it in our first Ca40 use case. (or do we?) These commented lines are a crude way for me to remember what the proper syntax of specifying such a nonlinear term would be.

File: test/test_ion_configurations.jl [hunk 14]
```diff
@@ -51,7 +51,7 @@ end
     @test any(isapprox.(linear_equilibrium_positions(10), pos, rtol=1e-4))
 
     # and test calculation of characterstic length scale for linear chain, equal mass
-    @test characteristic_length_scale(m_ca40, 1e6) ≈ 4.449042804354e-6
+    @test characteristic_length_scale(m_ca40, 1e6) ≈ 4.449042675683e-6
```
[the remark this answers is not in the document]
Comment (neil-glikin) [reply test/test_ion_configurations.jl:54]: The definition of the Ca40 mass changed slightly in the split-species branch. This change was made on 2020-07-02 in commit 4f0ef93.

**Pull request merged**